File: rust/day7_part1_2018.rs

use std::collections::{HashMap, HashSet};

fn main() {
    let input = std::fs::read_to_string("input.txt").unwrap();
    let mut dependencies: HashMap<char, HashSet<char>> = HashMap::new();
    let mut steps: HashSet<char> = HashSet::new();

    for line in input.lines() {
        let parts: Vec<char> = line.chars().collect();
        let step = parts[36];
        let dependency = parts[5];

        dependencies.entry(step).or_insert(HashSet::new()).insert(dependency);
        dependencies.entry(dependency).or_insert(HashSet::new());
        steps.insert(step);
        steps.insert(dependency);
    }

    let mut order: Vec<char> = Vec::new();

    while !steps.is_empty() {
        let mut available: Vec<char> = steps.iter().filter(|&step| dependencies[step].is_empty()).cloned().collect();
        available.sort();
        let next_step = available[0];
        order.push(next_step);
        steps.remove(&next_step);
        for deps in dependencies.values_mut() {
            deps.remove(&next_step);
        }
    }

    println!("{}", order.iter().collect::<String>());
}

File: r/day7_part1_2018.R

main <- function() {
    data <- readLines("input.txt")
    steps <- character(0)
    dependencies <- list()

    for (line in data) {
        step <- substr(line, 6, 6)
        next_step <- substr(line, 37, 37)
        steps <- c(steps, step, next_step)
        if (is.null(dependencies[[next_step]])) {
          dependencies[[next_step]] <- character(0)
        }
        dependencies[[next_step]] <- c(dependencies[[next_step]], step)
    }
    
    steps <- unique(steps)
    order <- character(0)

    while (length(steps) > 0) {
      available <- steps[sapply(steps, function(step) {
        is.null(dependencies[[step]]) || all(dependencies[[step]] %in% order)
      })]
      next_step <- min(available)
      order <- c(order, next_step)
      steps <- steps[steps != next_step]
    }

    cat(paste(order, collapse = ""))
}

main()
